src/celeritas/optical/OpticalTrackData.cc

```diff
@@ -0,0 +1,60 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/OpticalTrackData.cc
+//---------------------------------------------------------------------------//
+#include "OpticalTrackData.hh"
+
+#include "corecel/data/CollectionBuilder.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Resize states in host code.
+ */
+template<MemSpace M>
+void resize(OpticalStateData<Ownership::value, M>* state,
+            HostCRef<OpticalParamsData> const& params,
+            StreamId stream_id,
+            size_type size)
+{
+    CELER_EXPECT(params);
+    CELER_EXPECT(stream_id);
+    CELER_EXPECT(size > 0);
+    CELER_VALIDATE(stream_id < params.scalars.max_streams,
+                   << "multitasking stream_id=" << stream_id.unchecked_get()
+                   << " exceeds max_streams=" << params.scalars.max_streams);
+#if CELERITAS_CORE_GEO != CELERITAS_CORE_GEO_GEANT4
+    resize(&state->geometry, params.geometry, size);
+#else
+    // Geant4 state is stream-local
+    resize(&state->geometry, params.geometry, stream_id, size);
+#endif
+    CollectionBuilder{&state->materials}.resize(size);
```

Comment:
```suggestion
    resize(&state->materials, size);
```
But will the size actually be the number of volumes?

Reply:
If we need both the regular *and* optical material IDs when tracking photons, it would make more sense to have the matid -> optical ID indirection. Otherwise, I think it would be a little more efficient to directly store optical materials?